node/node-sdk.md

```diff
@@ -0,0 +1,610 @@
+      const current = progressEvent.currentTarget.response.length;
+      const progress = Math.floor(current / total * 100);
+      Context.current().heartbeat(progress);
+    },
+  });
+  return response.data;
+}
+```
+
+### Registration
+Activities are automatically registered by path when using the project initializer.
+
+If for any reason you need custom registration, you may do so using `worker.registerActivities`:
+
+```ts
+// worker/index.ts
+import * as impl from './implementation/greeter'; // Non-standard path
+
+worker.registerActivities({ '@activities/greeter': impl });
+```
+
+### Payloads
+We support the `DataConverter` interface as in [Java][java-data-converter].
+
+## Workflows
+Workflow code looks like vanilla TypeScript.
+Workflow implementation must export a `workflow` object which will be used as handler methods for workflow invocations, signals, and queries.
+
+Workflows run in the same NodeJS process as the worker and the activities.
+Each workflow instance runs in a separate [V8 isolate][v8-isolate], a completely isolated execution context including global variables.
+The Temporal SDK injects and removes references from the global object to ensure that code running in the isolate is fully deterministic.
+
+### Interface
+A workflow's interface is used for validating the implementation and generating a type safe `WorkflowClient` and `ChildWorkflow`.
+
+Workflow interfaces are inteded for the implementation and are written in sync or async form
+meaning a method could return `number` or it could return `Promise<number>` or their union.
+We provide a `WorkflowForClient` type which transforms the interface to an async only interface for `WorkflowClient` and `ChildWorkflow`.
+
+Workflow interface declarations are optional, they're only required for generating type safe clients.
+It is considered good practice to declare an interface for each workflow.
+
+```ts
+// interfaces/workflows.ts
+import { WorkflowForClient } from '@temporal-sdk/interfaces';
+
+// This interface will need to be implemented
+export interface Example {
+  main(name: string): Promise<void>;
+  // more on these later
+  signals: {
+    abandon(reason: string): void; // or Promise<void>; 
+  };
+  queries: {
+    step(): number;
+    state(key: string): unknown;
+  };
+}
+
+// Optionally export the client interface too
+export type GreeterClient = WorkflowClient<Example>;
+```
+
+### Implementation
+The workflow implementation is limited to using deterministic code only.
+A workflow file must export a `workflow` object which can be type checked using a pre-defined inteface.
+
+```ts
+// workflows/example.ts
+import '@temporal-sdk/workflow'; // inject global type declarations
+import { Example } from '@interfaces/workflows';
+
+async function main(name: string) {
+  // setTimeout and Math.random are replaced with deterministic versions
+  await new Promise((resolve) => setTimeout(resolve, Math.random()));
```

Comment:
This feels maybe a bit too magical to me. My initial inclination would be to change anything that'd be nondeterministic in the JS builtins to simply throw an error with a message saying "Please use `import { setTimeout } from '@temporal-sdk/determinism'` instead" (or similar) so that we have a bit more explicit control here, and don't have to explain to users why the built-ins work differently in a perhaps less-than-obvious way if they didn't fully read the docs.

Curious to hear everyone's thoughts on that topic.

Reply:
You have a point but I tend to think the opposite.
One of the great things about this SDK is that it really feels magical, you write code that looks like JS even though it runs differently.
I'd rather make the workflow runtime deterministic by default instead of letting users run anything non-deterministic or fail at runtime.

Reply:
I agree with @bergundy on this one. Our goal is to minimize nonstandard API surface. And standard random and time should work deterministically without the user caring about it.